test/cpp/end2end/concurrent_resumption_test.cc

```diff
@@ -0,0 +1,143 @@
+constexpr char kClientCertPath[] = "src/core/tsi/test_creds/client.pem";
+constexpr char kClientKeyPath[] = "src/core/tsi/test_creds/client.key";
+constexpr char kMessage[] = "Hello";
+
+std::string ReadFile(const char* file_path) {
+  grpc_slice slice;
+  GPR_ASSERT(
+      GRPC_LOG_IF_ERROR("load_file", grpc_load_file(file_path, 0, &slice)));
+  std::string file_contents(grpc_core::StringViewFromSlice(slice));
+  grpc_slice_unref(slice);
+  return file_contents;
+}
+
+class GrpcResumptionTest : public ::testing::Test {
+ protected:
+  void RunServer() {
+    std::string root_cert = ReadFile(kCaCertPath);
+    grpc::SslServerCredentialsOptions::PemKeyCertPair key_cert_pair = {
+        ReadFile(kServerKeyPath), ReadFile(kServerCertPath)};
+    grpc::SslServerCredentialsOptions sslOpts;
+    sslOpts.pem_key_cert_pairs.push_back(key_cert_pair);
+    sslOpts.pem_root_certs = root_cert;
+    sslOpts.force_client_auth = true;
+
+    grpc::ServerBuilder builder;
+    TestServiceImpl service_;
+
+    builder.AddListeningPort(server_addr_, grpc::SslServerCredentials(sslOpts));
+    builder.RegisterService("foo.test.google.fr", &service_);
+    server_ = builder.BuildAndStart();
+    std::cout << "GREG: server waiting on " << server_addr_ << "\n";
+    server_->Wait();
+  }
+
+  //   void TearDown() override {
+  //     if (server_ != nullptr) {
+  //       server_->Shutdown();
+  //       server_thread_->join();
+  //     }
+  //     server_thread_ = nullptr;
+  //   }
+
+  TestServiceImpl service_;
+  std::unique_ptr<Server> server_;
+  std::thread* server_thread_;
+  std::string server_addr_;
+};
+
+TEST_F(GrpcResumptionTest, ConcurrentResumption) {
+  int port = grpc_pick_unused_port_or_die();
+  std::ostringstream addr;
+  addr << "localhost:" << port;
+  server_addr_ = addr.str();
+  std::cout << "GREG: Serving on " << server_addr_ << "\n";
+  server_thread_ = new std::thread([&]() { RunServer(); });
+  sleep(1);
+
+  std::string root_cert = ReadFile(kCaCertPath);
+  std::string client_key = ReadFile(kClientKeyPath);
+  std::string client_cert = ReadFile(kClientCertPath);
+  grpc::SslCredentialsOptions sslOpts;
+  sslOpts.pem_root_certs = root_cert;
+  sslOpts.pem_private_key = client_key;
+  sslOpts.pem_cert_chain = client_cert;
+
+  grpc_ssl_session_cache* cache = grpc_ssl_session_cache_create_lru(16);
+  ChannelArguments channel_args;
+  channel_args.SetPointer(std::string(GRPC_SSL_SESSION_CACHE_ARG), cache);
+  channel_args.SetSslTargetNameOverride("foo.test.google.fr");
+
+  std::shared_ptr<Channel> channel = grpc::CreateCustomChannel(
+      server_addr_, grpc::SslCredentials(sslOpts), channel_args);
+
+  auto stub = grpc::testing::EchoTestService::NewStub(channel);
+  grpc::testing::EchoRequest request;
+  grpc::testing::EchoResponse* response = nullptr;
+  request.set_message(kMessage);
+  ClientContext context;
+  context.set_deadline(grpc_timeout_milliseconds_to_deadline(100));
```

Comment:
To reduce some flakiness that I noticed while testing, I would bump this up to 1000.